editor/src/clj/editor/field_expression.clj

```diff
@@ -33,21 +34,33 @@
       (math/round-with-precision (op a b) precision))
     (parse-fn (parsable-number-format text))))
 
-(defn to-int [s]
+(defn- to-double-safe [s]
+  (try
+    (-> (ExpressionBuilder. s) .build .evaluate)
+    (catch Throwable _
+      nil)))
+
+(defn- to-int-legacy [s]
   (try
     (int (evaluate-expression #(Integer/parseInt %) 1.0 s))
     (catch Throwable _
       nil)))
 
+(defn- to-double-legacy [s]
+  (try
+    (evaluate-expression #(Double/parseDouble %) math/precision-general s)
+    (catch Throwable _
+      nil)))
+
+(defn to-int [s]
+  (or (int (to-double-safe s)) (to-int-legacy s)))
+
 (defn format-int
   ^String [n]
   (if (nil? n) "" (str n)))
 
 (defn to-double [s]
- (try
-   (evaluate-expression #(Double/parseDouble %) math/precision-general s)
-   (catch Throwable _
-     nil)))
+  (or (to-double-safe s) (to-double-legacy s)))
```

Comment:
I agree we should probably get rid of the legacy way of doing it, but perhaps we should only change the order we attempt this so that we try to call `Double/parseDouble` or `Integer/parseInt` first, and if they throw an exception we fall back on the `ExpressionBuilder`?

I think that would allow us to support the special sentinel values like `NaN` and `Infinity` for doubles. People can do whatever they want in their game projects, so potentially `Infinity` could have a special meaning in their game logic and they need to be able to have it as a value. I think it is fine to not allow them in arithmetic expressions, though.

Reply:
I updated.